tools/pubsys/src/aws/validate_ami/mod.rs

```diff
@@ -0,0 +1,847 @@
+
+/// Validates EC2 images by calling `describe-images` on all images in the file given by
+/// `expected-amis-path` and ensuring that the returned `public`, `ena-support`,
+/// `sriov-net-support`, and `launch-permissions` fields have the expected values.
+#[derive(Debug, StructOpt)]
+#[structopt(setting = clap::AppSettings::DeriveDisplayOrder)]
+pub(crate) struct ValidateAmiArgs {
+    /// File holding the expected amis
+    #[structopt(long, parse(from_os_str))]
+    expected_amis_path: PathBuf,
+
+    /// Optional path where the validation results should be written
+    #[structopt(long, parse(from_os_str))]
+    write_results_path: Option<PathBuf>,
+
+    #[structopt(long, requires = "write-results-path")]
+    /// Optional filter to only write validation results with these statuses to the above path
+    /// The available statuses are: `Correct`, `Incorrect`, `Missing`.
+    write_results_filter: Option<Vec<AmiValidationResultStatus>>,
+
+    #[structopt(long)]
+    /// If this argument is given, print the validation results summary as a JSON object instead
+    /// of a plaintext table
+    json: bool,
+}
+
+/// Performs EC2 image validation and returns the `AmiValidationResults` object
+pub(crate) async fn validate(
+    args: &Args,
+    validate_ami_args: &ValidateAmiArgs,
+) -> Result<AmiValidationResults> {
+    info!("Parsing Infra.toml file");
+
+    // If a lock file exists, use that, otherwise use Infra.toml
+    let infra_config = InfraConfig::from_path_or_lock(&args.infra_config_path, false)
+        .context(error::ConfigSnafu)?;
+
+    trace!("Parsed infra config: {:#?}", infra_config);
+
+    let aws = infra_config.aws.unwrap_or_default();
+
+    // Parse the expected ami file
+    info!("Parsing expected ami file");
+    let expected_images = parse_expected_amis(&validate_ami_args.expected_amis_path).await?;
+
+    info!("Parsed expected ami file");
+
+    // Create a `HashMap` of `AmiClient`s, one for each region where validation should happen
+    let base_region = &Region::new(
+        aws.regions
+            .get(0)
+            .ok_or(error::Error::EmptyInfraRegions {
+                path: args.infra_config_path.clone(),
+            })?
+            .clone(),
+    );
+    let mut ami_clients = HashMap::with_capacity(expected_images.len());
+
+    for region in expected_images.keys() {
+        let client_config = build_client_config(region, base_region, &aws).await;
+        let ami_client = AmiClient::new(&client_config);
+        ami_clients.insert(region.clone(), ami_client);
+    }
+
+    // Retrieve the EC2 images using the `AmiClient`s
+    info!("Retrieving EC2 images");
+    let images = describe_images(&ami_clients, &expected_images)
+        .await
+        .into_iter()
+        .map(|(region, result)| {
+            (
+                region,
+                match result {
+                    Ok(result) => Ok(result),
+                    Err(_) => Err(error::Error::UnreachableRegion {
```

Comment:
nit/tip (don't feel compelled to change this): I think this is the same as

```rust
result.map_err(|_| error::Error::UnreachableRegion { region: region.to_string() })
```

`match` on `Result` and `Option` when the operation to be carried out is already expressed as a method on those structs tends to result in deep nesting.